Feature/lab2 steps (#18)
ArctiqTeam/arctiq-copilot-trail: labs/lab2-development-use-cases/README.md

- Explaining code and errors
- Writing documentation
- Adding a feature to the codebase

While learning about doing these things we will talk about improving the output of Copilot both by applying the prompt crafting methods previously discussed and the best-practices when encountering issues like:

- missing or changed elements of code
- hallucinations
- incomplete or naive implementations
- stalling and repetitive answers

## Step 1 - Common Problems and Solutions

First, a little more theory. Based on the prompts you have crafted and fed into Copilot, or the amount of code you are working with, you may or may not get the results you wanted. There are a number of ways in which Copilot's answers can fall short of expectations. Commonly you may see it:

- stalls on autocomplete
- repeats itself
- deviates from existing code in answer
- omits or hallucinates content
- provides an obsolete or dated answer
- gives a naive or inefficient solution
- retracts an answer that matches public code

### Copilot Stalls on Autocomplete

> - It can happen that Copilot does not start or provide an answer when it seems clear to the user that they are pausing for a suggestion. This happens most often in the inline copilot suggester. The simple fix here is to give it a little push by starting either a comment or a method and it should trigger copilot.

Author: good catch. I like the suggestion

### Copilot Repeats Itself

> - It can also happen that Copilot will regenerate the same content over and over again as if it has forgotten that it just gave that code. This is a form of hallucination error and does illustrate why copilot is not in the driver's seat. It requires an experienced user to scrutinize the suggested code and decide if it should be incorporated into the codebase or if more context is required.

Author: We can mention this in the session, but I will leave that example out for now.

### Gives a Naive or Inefficient Answer

- Copilot may answer a tricky question with a simple, effective, but inefficient answer. It often finds an answer that meets the objectives but if efficiency isn't a stated goal in the prompt, it may not prioritize that and the developer will need to re-request that the answer be optimized in the way they desire. The lesson is that you cannot assume Copilot will always choose the optimal solution-- it's goal is to meet the prompt and if efficiency isn't specified it might settle on the simplest answer.

### Retracts an Answer that Matches Public Code

- One setting in your GitHub account or Organization is the toggle that Copilot will filter results that match public code. This means that if it generates an answer that happens to match a public code solution for the same thing it will retract the answer. This can be very frustrating and actually fairly common. The simple solution to this is to add to the prompt the phrase "Do not match public code." This informs Copilot that it should attempt to originate a new answer specifically avoiding a public code match. It really works.

> Keep these practices in mind when submitting prompts to Copilot. You will see many of these issues.

Reviewer: One other thing that I do is say "output only the minimum amount of code needed to do this change" and it fixes the public code match even though it regenerates the same code.

Author: that's an interesting prompt, but it kinda sounds like a hack. We can mention that in the class, but I will leave it from the standard reference for now.

- You can, of course, prompt from an inline location or using the Copilot Chat sidebar. This can be helpful if the program logic is not very clear or unnecessarily convoluted.

    ![Prompt to explain a method inline in JetBrains Rider](images/cs-app-inline-prompt-explain-rider.png)

    Figure 2: Prompting to explain a method inline in Rider

- The `/explain` keyword can be used to help understand cryptic error messages returned by the runtime or the build process. In VSCode the Copilot Chat can be directed to access `@terminal` with the keyword `/explain` to review the error output.

    ![Prompt to the VSCode terminal for explanation of a build warning](images/cs-app-vscode-terminal-explain.png)

    Figure 3: Requesting an explanation in VSCode to a build warning in the terminal.

- In IDEs that do not support the `@terminal` agent, you can cut-and-paste error messages into the chat window to include as part of the prompt.

    ![Prompt to JetBrains Rider for explanation of a build warning](images/cs-app-rider-build-explain.png)

    Figure 4: Requesting an explanation in JetBrains Rider to a build warning.

## Step 3 - Writing Documentation

> - In terms of documentation, Copilot can relieve a lot of tedium. Copilot can generate documenation for you and it can adopt your particular style of documentation given enough examples to work from. So write a few comments yourself and Copilot will resume in a consistent way. The special command for this activity is `/doc`.

Reviewer: It means the same but I've never seen somebody use "tedium" before.
```suggestion
- In terms of documentation, Copilot can relieve a lot of tediousness. Copilot can generate documentation for you and it can adopt your particular style of documentation given enough examples to work from. So write a few comments yourself and Copilot will resume in a consistent way. The special command for this activity is `/doc`.
```

- The following is an example of a Mermaid diagram that Copilot generated for the chess Web Application:

    ```mermaid
    erDiagram
    GameState {
        Board Board
        PieceColor CurrentTurn
        bool IsGameOver
        Position LastMoveFrom
        Position LastMoveTo
        string FEN
    }

    Board {
        Piece Squares[8x8]
    }

    Piece {
        PieceColor Color
        Position Position
        PieceType Type
        string ImagePath
        char Symbol
    }

    Pawn {
        char Symbol
    }

    Queen {
        char Symbol
    }

    Position {
        int Row
        int Column
    }

    GameState ||--|| Board: contains
    Board ||--o{ Piece: contains
    Piece ||--|| Position: has
    Pawn ||--|| Piece: inherits
    Queen ||--|| Piece: inherits
    ```

## Step 4 - Adding a Feature to the Codebase

### The ChessWeb Codebase

- Our example application is called ChessWeb. It is a simple Web Application designed to have just enough complexity in its function to be an interesting target for Copilot work. It can be found in the repository under chessweb-cs (C#), chessweb-node (NodeJS), and chessweb-java (Java).

> - This sample application is far from a complete Chess-playing interface. In fact, all it does is to provide a view of a Chess board that allows turn-based moves to be entered with very basic piece rules applied. It serves as a basis for extension using Copilot. If you don't have knowledge of or interest in Chess, don't worry, no expertise is required to participate in this part.

Reviewer: ```suggestion
- This sample application is far from a complete Chess-playing interface. In fact, all it does is to provide a view of a chessboard that allows turn-based moves to be entered with very basic piece rules applied. It serves as a basis for extension using Copilot. If you don't have knowledge of or interest in Chess, don't worry, no expertise is required to participate in this part.
```

- Let's take a look at what we have:

- We can start the application by entering the `chessweb-cs` directory and running it.

    ```bash
    cd chessweb-cs
    dotnet run
    ```

- If successful we can now open a browser to the URI `http://localhost:5111` and see something like this:

    ![Application Startup View](images/cs-app-view-startup.png)

    Figure 8: Application Startup View

> - In this view we can see a familiar Chess board with pieces set up for the beginning of a game. There is a button that can be used to reset the position. The board is labeled on the ranks and files: ranks are labeled 1-8 and files are labeled a-h. Finally, there is a turn indicator to the right of the board that will display which side and which colour is due to play.

Reviewer: ```suggestion
- In this view we can see a familiar chessboard with pieces set up for the beginning of a game. There is a button that can be used to reset the position. The board is labeled on the ranks and files: ranks are labeled 1-8 and files are labeled a-h. Finally, there is a turn indicator to the right of the board that will display which side and which colour is due to play.
```

### Extending the Game State Display

- A reasonable feature to add is a display of the current board position in a standard text-based representation known as Forsyth-Edwrads Notation or FEN. This is a text encoding that can be used to rebuild a Chess position and represent the current game state in both a human and machine-readable fashion. You can read more about FEN [here](https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation).

- I will give Copilot a suitable prompt to begin this change:

    ![add-FEN-box-prompt](images/cs-app-add-fen-box-prompt.png)

    Figure 9: Prompting for a FEN display box

    ```prompt
    Let's extend this interface to display the current board position as 
    Forsyth-Edwards Notation (FEN). This should take the form of a text 

>     input box below the chess board with the current FEN string of the board 

Reviewer: ```suggestion
    input box below the chessboard with the current FEN string of the board 
```
